Code revision in diff.py

mcara opened the pull request:
This PR cleans up the code a little and moves constant (relative to the loop) operations out of the loop. See https://github.com/astropy/astropy/issues/7468 and the links provided there.

EDIT: Fix #7468 . Follow up of #7444

saimn commented on astropy/utils/diff.py:
```diff
@@ -99,27 +91,30 @@ def report_diff_values(a, b, fileobj=sys.stdout, indent_width=0):
 
-    padding = max(len(typea.__name__), len(typeb.__name__)) + 3
+    typea = type(a)
+    typeb = type(b)
```
Here `a` and `b` have been converted to string, so this is not the same implementation as before, maybe the cause of the failure?

mcara commented:
Yeah, I've noticed this mistake (of moving type determination) but was offline to fix it.

pllim commented:
Still fails with the same error. FYI. https://travis-ci.org/astropy/astropy/jobs/379478016

mcara commented:
@pllim @saimn I know... working on it. There are two issues:

1. Why is `repr(of string)` necessary here: https://github.com/astropy/astropy/blob/e80c6c44ea1e3c1b9154f3d34e523db1d5304ec1/astropy/utils/diff.py#L74

2. [most importantly] This and comment above it, seems to be a complete nonsense:

```python
In [2]: a = np.array([[1,2,3,4,7],[5,6,7,8,7],[1,1,1,1,1]])

In [3]: b = np.array([[1,2,3,0,7],[5,6,0,8,7],[1,1,1,1,1]])

In [4]: diff.report_diff_values(a, b, fileobj=sys.stdout)
  at [0, 3]:
line: 'a> 4'
    a> 4
line: 'b> 0'
    b> 0
  at [1, 2]:
line: 'a> 7'
    a> 7
line: 'b> 0'
    b> 0
  ...and at 1 more indices.
Out[4]: False
```

There are ONLY TWO differences between `a` and `b` above!

mcara commented:
@pllim @saimn I know... working on it. There are two issues:

1. Why is `repr(of string)` necessary here: https://github.com/astropy/astropy/blob/e80c6c44ea1e3c1b9154f3d34e523db1d5304ec1/astropy/utils/diff.py#L74

2. [most importantly] This https://github.com/astropy/astropy/blob/e80c6c44ea1e3c1b9154f3d34e523db1d5304ec1/astropy/utils/diff.py#L88 and comment above it, seems to be a complete nonsense:

```python
In [2]: a = np.array([[1,2,3,4,7],[5,6,7,8,7],[1,1,1,1,1]])

In [3]: b = np.array([[1,2,3,0,7],[5,6,0,8,7],[1,1,1,1,1]])

In [4]: diff.report_diff_values(a, b, fileobj=sys.stdout)
  at [0, 3]:
line: 'a> 4'
    a> 4
line: 'b> 0'
    b> 0
  at [1, 2]:
line: 'a> 7'
    a> 7
line: 'b> 0'
    b> 0
  ...and at 1 more indices.
Out[4]: False
```

There are ONLY TWO differences between `a` and `b` above!

pllim commented:
I don't know the history of why it was written that way. This code came from `io.fits`, which in turn came from PyFITS. The "complete nonsense" is my bad; I wrote the comment in an attempt to understand the logic and forgot to remove it. Feel free to improve the comment as you see fit.

mcara commented:
No, the problem is not in the comment - comment just tries to explain the code. I do not understand how there can be 625 different values when a `5x5` array has only 25 values in total!

pllim commented:
> how there can be 625 different values when a 5x5 array has only 25 values in total!

Each 5x5 array has 25 elements in total. And that `reduce` thingy appears to perform `25 * 25` when I ran the logic manually, hence results in `num_diffs = 625` when all elements do not match. But I cannot explain why...

mcara commented:
But then the rest of the logic does not work when, e.g., there are only 2 different indices in `diff_indices` such as in my example. Then code should print 2 indices but it also thinks that it already printed 3 and reports that there is "one more".

pllim commented:
Yeah. I wonder if it shouldn't just be `num_diffs = len(diff_indices[0])`. Even with n-D arrays, that would still work, right?

mcara commented:
I think so.

pllim commented:
I dug a little and it seems to be a `lambda` function prior to #5247 that would still give the same 625 answer. Before that, its history was lost in repo transition. Only @embray knows, if he remembers... :pray: 
```python
num_diffs = reduce(lambda x, y: x * y, (len(d) for d in diff_indices), 1)
```

mcara commented:
In my last commit I restored the logic of using `repr` on strings and on numeric types (but use `str` on _other_ types) from the original implementation. It is not clear to me why this is necessary (i.e., how is `str(1)` different from `repr(1)`?)

I also modified the logic for counting the number of different elements between `ndarrays`. By the way, I do not see how this function can work if arrays have a different number of elements. It should just return `False` but currently, I think, it will crash.

Finally, I have modified the logic when precomputing values in order to minimize the number of `isinstance()` calls, which, in my experience, are slow.

pllim commented:
>  how is str(1) different from repr(1)

Keep in mind that this is heavily used in `fitsdiff` (header, image, table, what-nots) . There is a chance somewhere in the FITS world, `__str__` and `__repr__` give different outputs.

mcara commented:
Well, `repr()` is used only on 1. numeric types (`int`, `float`, `complex`, `np.number`) and 2. sometimes on string types (but only when one argument is `str` and the other one is not). The logic of this is escaping me.

pllim commented on astropy/io/fits/tests/test_diff.py:
```diff
@@ -554,7 +554,7 @@ def test_different_table_data(self):
                 '    a> True\n'
                 '    b> False') in report
-        assert ('...and at 13 more indices.\n'
+        assert ('...and at 1 more indices.\n'
```
I am a bit concerned in the change of this number. `13 -> 1` does not seem correct here given that "13" is still tested for in two other places in this same test. Perhaps we should take a step back and modify the code in such a way that it does not fundamentally change that `fitsdiff` does...

mcara replied:
This change can be explained: this is an array with 4 elements (`2x2`) and so there are 4 differing elements. In the past `fitsdiff` thought that a `2x2` array would have 16 different elements. `fitsdiff` displays first three differences so that in the new version (that correctly, _IMO_, computes 4 elements) there is "1 more" different element (that was not displayed) while in the old version (the one that was computing 16 differing elements for a `2x2` array) after displaying first three different elements there were "13 more" left.

I do not thing we need to take a step back. Instead, I would challenge you to display all 16 different elements in the old code. That is, in `islice` do not limit the code to take the first three elements - print them all and see if you get 16 different elements.

mcara replied:
The other 13 in the test comes form the fact that there are `fitsdiff` has no idea how to count things (I left that one for you @pllim :-) - well it needs some discussion/understanding of what was the intension). Take a look at the "report":

mcara replied:
... or, is this just a bad test setup?! Column C is the only one that has 4 "rows" (array data of shape `2x4`) while all other columns have only two rows (array data of shape Nx2)!

mcara replied:
... or, maybe, it's not a bad test but just how things are supposed to be for this particular `dim` parameter. I just do not know what is going on here. In any case, the two 13th reported in the original test are unrelated.

saimn replied:
> Let's consider "Column G" for example. This one contains this array array=[0.0, 1.0] (which has a single row (in my interpretation) with two columns)

No, it's a column, with 2 rows.

Maybe easier to see with the table:
```
In [8]: ta.data
Out[8]: 
FITS_rec([( 1, [  0], [[0, 1], [2, 3]], 0, 'abc', 0., 0., 0.+1.j, 4.+5.j, [0, 1]),
          ( 0, [  1], [[4, 5], [6, 7]], 2, 'def', 1., 1., 2.+3.j, 6.+7.j, [2, 3])],
         dtype=(numpy.record, [('A', 'i1'), ('B', 'u1', (1,)), ('C', '<i2', (2, 2)), ('D', '<i4'), ('E', 'S3'), ('F', '<f4'), ('G', '<f8'), ('H', '<c8'), ('I', '<c16'), ('J', '<i4', (2,))]))

In [9]: Table(ta.data)
Out[9]: 
<Table length=2>
  A   B [1] C [2,2]   D    E      F       G        H         I        J   
 bool  bool  int16  int32 str3 float32 float64 complex64 complex128 object
----- ----- ------- ----- ---- ------- ------- --------- ---------- ------
 True False  0 .. 3     0  abc     0.0     0.0        1j     (4+5j)  [0 1]
False  True  4 .. 7     2  def     1.0     1.0    (2+3j)     (6+7j)  [2 3]
```

> For C there is a single row with 4 columns in the array (array=[[0, 1, 2, 3], [4, 5, 6, 7]])

Because of `dim`, C is a special beast here. Each row element is a 2x2 array.

So, in my understanding the report looks correct (once you know what's in the table).

mcara replied:
@saimn Thanks for clarifying this! I think I now understand what is going on with C. In this case the reported 13 differences in the table look correct and, IMO, the reported "... 1 more" for the row 1 in column C also seems correct. I think we should not revert to the old logic of counting differences by which a `5x5` array can have 625 differences.

saimn replied:
Agreed, I don't understand either the old `num_diffs` logic, and the [history](https://github.com/spacetelescope/PyFITS/commit/8e4d6d20e0fddba232f88ce3edf985b871816910#diff-ec63a1f7dd27e3295d30eb67aabc6e79L1080) of it does not help much.

pllim replied:
So, what does this mean? There needs to be two different logic when dealing with columns and when dealing with Numpy arrays?

mcara replied:
On the contrary, now in both cases the same logic of counting differences is used. That is, now when counting arrays you get the _actual_ number of differences - not some unjustifiable number.

pllim replied:
Yes. I think you are right. I'll need to wrap my head around this but not right now...

embray replied:
Yes, the old formatting for this absolutely sucked.  The use of the diff format is useful for string values, but not for numerical differences.

saimn commented:
> how is str(1) different from repr(1)

`repr` has additional quotes around the string. So maybe it is needed when comparing to a non-string value which is also converted to string with `repr`.

mcara commented:
@pllim 

> @mcara , can you please rebase to pick up fix for Sphinx build? Thanks.

Now this is what truly makes this an expert-level issue. 😅

mcara commented:
@pllim Done rebasing.

mcara commented:
Why is this PR stuck?

pllim commented:
@mcara , as discussed offline, I'll get to this after some other higher priority tasks. Thanks for your patience.

mcara commented on astropy/utils/diff.py:
```diff
@@ -67,59 +66,72 @@ def report_diff_values(a, b, fileobj=sys.stdout, indent_width=0):
+        lnpad = ' '
+        sign_a = 'a>'
+        sign_b = 'b>'
+        if isinstance(a, _NUMERIC_TYPES):
```
@pllim Good point. Maybe we can switch to `numbers.Number` but see https://stackoverflow.com/questions/44756406/python-number-base-class-or-how-to-determine-a-value-is-a-number?#comment76494404_44756520 If you believe compatibility is not an issue - we can switch.

pllim replied:
Travis CI currently tests back to Numpy 1.11. So if tests pass with this switch, we can ensure compatibility with `numpy>=1.11`, no? If not, perhaps `isinstance(a, (numbers.Number, np.number))` would cover everything?

mcara commented on astropy/utils/tests/test_diff.py:
```diff
@@ -74,7 +74,23 @@ def test_array_comparison():
+    """
+    f = io.StringIO()
+    a = np.empty((1, 2, 3))
+    b = np.empty((2, 3))
```
@pllim We do need `b`. The point of this check is to compare, detect, and report when two `numpy` arrays have different shapes (hence the name of the function). `https://github.com/mcara/astropy/blob/22352e4b490fd95566d5780a6cf71130ebd2db5c/astropy/utils/diff.py#L70` is performed only when both variables are `numpy.ndarray`s.

pllim replied:
@mcara , I don't understand. `b = a[0]` has a different shape and is *still* an `ndarray`, so why won't that work?

mcara replied:
@pllim Oh, I was thinking about something else. Yes, this should work.

pllim commented:
For the sake of CircleCI, can you please do a rebase (and also `git submodule update` if needed) against `master`? Thanks!

pllim commented:
@astrofrog or @larrybradley , do you know why CircleCI failed?

pllim commented on astropy/utils/tests/test_diff.py:
```diff
@@ -54,6 +54,12 @@ def test_diff_types():
+    f = io.StringIO()
+    assert not report_diff_values(1.0, 1, fileobj=f)
+    out = f.getvalue()
+    assert out.lstrip('u') == '  (float) a> 1.0\n    (int) b> 1\n'
```
Why does this one test need `.lstrip('u')`? I don't see it used in the other tests.

mcara replied:
@pllim I just copied _your own_ code from a previous function: https://github.com/astropy/astropy/pull/7444/files#diff-e98489b2b61bd9ec323e65176107ed37R53

mcara replied:
@pllim Do you know did you put `lstrip()` there?

pllim replied:
@mcara , mine was a copy-and-paste oversight from `io.fits`. Please remove that one too. Thank you for your *continued* patience! :wink:

pllim commented:
Test failure is unrelated and is being handled by #7544. Mihai, you can go pester Larry down the hall about it if you want. :wink:
